arwen/contexts/runtime.go

```diff
@@ -572,40 +576,76 @@ func (context *runtimeContext) ExecuteAsyncCall(address []byte, data []byte, val
 	return nil
 }
 
-func (context *runtimeContext) SetAsyncCallInfo(asyncCallInfo *arwen.AsyncCallInfo) {
-	context.asyncCallInfo = asyncCallInfo
+func (context *runtimeContext) CreateAndAddAsyncCall(
+	groupID []byte,
+	address []byte,
+	data []byte,
+	value []byte,
+	successCallback []byte,
+	errorCallback []byte,
+	gas uint64,
+) error {
+	metering := context.host.Metering()
+	err := metering.UseGasForAsyncStep()
+
+	gasToLock := uint64(0)
+	shouldLockGas := context.HasCallbackMethod() || !context.host.IsDynamicGasLockingEnabled()
```

Comment:
Intention:
* If dynamic gas locking has not been enabled yet, then always lock gas.
* If dynamic gas locking has been enabled, then only lock gas if the contract has a callback method, otherwise there's no need.

Is it ok?

Reply:
It is ok, but I would treat backward compatibility separately thus if !enabled -> do legacy stuff, otherwise do new stuff which checks for callback function existence and so.

Reply:
Refactored for clarity.